substrate/tendermint/client/src/authority/mod.rs

```diff
@@ -0,0 +1,453 @@
+            };
+
+            // There's a race condition between the machine add_block and this
+            // Both wait for a write lock on this number and don't release it until after updating
+            // it accordingly
+            {
+              let mut number = number.write().unwrap();
+              if this_number < *number {
+                continue;
+              }
+              let new_number = this_number + 1;
+              *number = new_number;
+              recv = gossip.messages_for(TendermintGossip::<T>::topic(new_number));
+            }
+
+            let justifications = import.client.justifications(notif.hash).unwrap().unwrap();
+            step.send((
+              BlockNumber(this_number),
+              Commit::decode(&mut justifications.get(CONSENSUS_ID).unwrap().as_ref()).unwrap(),
+              // This will fail if syncing occurs radically faster than machine stepping takes
+              // TODO: Set true when initial syncing
+              get_proposal(&env, &import, &notif.header, false).await
+            )).await.unwrap();
+          } else {
+            break;
+          }
+        },
+
+        // Machine reached a new height
+        new_number = new_number_rx.next() => {
+          if new_number.is_some() {
+            recv = gossip.messages_for(TendermintGossip::<T>::topic(*number.read().unwrap()));
+          } else {
+            break;
+          }
+        },
+
+        // Message to broadcast
+        msg = gossip_rx.next() => {
+          if let Some(msg) = msg {
+            let topic = TendermintGossip::<T>::topic(msg.number().0);
+            gossip.gossip_message(topic, msg.encode(), false);
+          } else {
+            break;
+          }
+        },
+
+        // Received a message
+        msg = recv.next() => {
+          if let Some(msg) = msg {
+            messages.send(
+              match SignedMessage::decode(&mut msg.message.as_ref()) {
+                Ok(msg) => msg,
+                Err(e) => {
+                  // This is guaranteed to be valid thanks to to the gossip validator, assuming
+                  // that pipeline is correct. This doesn't panic as a hedge
+                  error!(target: "tendermint", "Couldn't decode valid message: {}", e);
+                  continue;
+                }
+              }
+            ).await.unwrap();
+          } else {
+            break;
+          }
+        }
+      }
+    }
+  }
+}
+
+#[async_trait]
+impl<T: TendermintValidator> Network for TendermintAuthority<T> {
+  type ValidatorId = u16;
+  type SignatureScheme = TendermintValidators<T>;
+  type Weights = TendermintValidators<T>;
+  type Block = T::Block;
+
+  const BLOCK_PROCESSING_TIME: u32 = T::BLOCK_PROCESSING_TIME_IN_SECONDS;
+  const LATENCY_TIME: u32 = T::LATENCY_TIME_IN_SECONDS;
+
+  fn signer(&self) -> TendermintSigner<T> {
+    self.active.as_ref().unwrap().signer.clone()
+  }
+
+  fn signature_scheme(&self) -> TendermintValidators<T> {
```

Comment:
Considering it's more than just the party, yet also signature aggregation and verification, I do believe it's a scheme. It is a contextual scheme however...

Reply:
Resolving since this is low impact and we have no better name clearly available. It may theoretically be worth merging with weights in the future but...